### PR TITLE
fix: cancel background stream job if table fragment not persist

### LINKED ISSUE #14104: Recovery takes a long time in deterministic test
I recently saw in https://buildkite.com/risingwavelabs/main-cron/builds/1372#018c87f2-5dac-4891-a103-73b931341efd that `recovery start` can take a long time to happen.

After nodes killed at ~23:34, after about 2.5hrs at 02:05, recovery still doesn't start, which is strange to me. It somehow makes the background mv to be stuck as well.

```
2022-10-28T23:34:15.330083Z  INFO risingwave_simulation::cluster: kill frontend-1
2022-10-28T23:34:15.520903Z  INFO risingwave_simulation::cluster: kill compute-2
2022-10-28T23:34:15.567393Z  INFO risingwave_simulation::cluster: kill compute-1
2022-10-28T23:34:15.632018Z  INFO risingwave_simulation::cluster: kill frontend-2
2022-10-28T23:34:15.842152Z  INFO risingwave_simulation::cluster: kill compute-3

2022-10-28T23:34:16.267249Z  INFO failure_recovery{err=gRPC request failed: Client specified an invalid argument: 
to collect actors not found prev_epoch=3261389423312896}: 
risingwave_meta::barrier::recovery: recovery start!
2022-10-28T23:34:39.028666Z  INFO failure_recovery{err=gRPC request failed: Client specified an invalid argument:
 to collect actors not found prev_epoch=3261389423312896}: risingwave_meta::barrier::recovery: 
recovery success epoch=3261391109029888 paused=None

2022-10-28T23:34:50.660644Z DEBUG risingwave_simulation::slt: retry count iteration=0
2022-10-28T23:34:50.784392Z DEBUG risingwave_simulation::slt: Retry for background ddl iteration=0
2022-10-28T23:34:50.810071Z  INFO risingwave_simulation::cluster: kill compute-3
2022-10-28T23:34:50.951084Z  INFO risingwave_simulation::cluster: kill compactor-2
2022-10-28T23:34:51.169632Z  INFO risingwave_simulation::cluster: kill compactor-1
2022-10-28T23:34:51.505341Z  INFO risingwave_simulation::cluster: kill compute-2
2022-10-29T00:04:50.833313Z ERROR risingwave_simulation::slt: failed to wait for background mv to finish creating iteration=0 err=failed to wait for background mv to finish creating for mv
2022-10-29T00:04:50.833313Z DEBUG risingwave_simulation::slt: retry count iteration=1
2022-10-29T00:04:52.850341Z DEBUG risingwave_simulation::slt: Retry for background ddl iteration=1 name="mv"
2022-10-29T00:34:52.896505Z ERROR risingwave_simulation::slt: failed to wait for background mv to finish creating iteration=1 err=failed to wait for background mv to finish creating for mv
2022-10-29T00:34:52.896505Z DEBUG risingwave_simulation::slt: retry count iteration=2
2022-10-29T00:34:56.910407Z DEBUG risingwave_simulation::slt: Retry for background ddl iteration=2 name="mv"
2022-10-29T01:04:56.947553Z ERROR risingwave_simulation::slt: failed to wait for background mv to finish creating iteration=2 err=failed to wait for background mv to finish creating for mv
2022-10-29T01:04:56.947553Z DEBUG risingwave_simulation::slt: retry count iteration=3
2022-10-29T01:05:04.962505Z DEBUG risingwave_simulation::slt: Retry for background ddl iteration=3 name="mv"
2022-10-29T01:35:05.010748Z ERROR risingwave_simulation::slt: failed to wait for background mv to finish creating iteration=3 err=failed to wait for background mv to finish creating for mv
2022-10-29T01:35:05.010748Z DEBUG risingwave_simulation::slt: retry count iteration=4
2022-10-29T01:35:21.031668Z DEBUG risingwave_simulation::slt: Retry for background ddl iteration=4 name="mv"
2022-10-29T02:05:21.071181Z ERROR risingwave_simulation::slt: failed to wait for background mv to finish creating iteration=4 err=failed to wait for background mv to finish creating for mv
2022-10-29T02:05:21.071181Z DEBUG risingwave_simulation::slt: retry count iteration=5
noelkwan@Noels-MacBook-Pro ~ %
```

shell cmd to generate these logs:
```
 cat ~/Downloads/recovery-streaming-11.log | rg '(kill|recovery start|recovery success|iteration|cluster is under recovering|backfill)'
```

### PR DESCRIPTION
**Pull request by yezizp2012**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Resolve #14104 

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/meta/src/rpc/ddl_controller.rs [hunk 1]
```diff
@@ -850,10 +850,18 @@ impl DdlController {
...
+                    tracing::error!(id = job_id, error = ?e, "finish stream job failed");
+                    if let Err(err) = self.fragment_manager.select_table_fragments_by_table_id(&job_id.into()).await
+                        && err.is_fragment_not_found() {
+                        // If the table fragments are not found, it means that the stream job has not been created.
+                        // We need to cancel the stream job.
+                        self.cancel_stream_job(&stream_job, internal_tables, Some(&e))
+                            .await?;
+                    } else {
+                        // NOTE: This assumes that we will trigger recovery,
+                        // and recover stream job progress.
```
Comment (kwannoel) [inline src/meta/src/rpc/ddl_controller.rs:863]: Shall we keep the trace here?

Comment (kwannoel) [reply src/meta/src/rpc/ddl_controller.rs:863]: Nevermind, it's already done above.

**Review by kwannoel — APPROVED** [review]

**Review by fuyufjh — APPROVED** [review]
Cool. Thanks a lot to @yezizp2012 @kwannoel

**Comment by yezizp2012** [comment]
Sadly it's not the root cause of #9712 , still need to investigate.

**Comment by fuyufjh** [comment]
> Sadly it's not the root cause of #9712 , still need to investigate.

Yeah, I was just about to ask why this can cause a freeze of CN🤣

File: src/meta/src/rpc/ddl_controller.rs [hunk 2]
```diff
@@ -850,10 +850,18 @@ impl DdlController {
...
-                // and recover stream job progress.
                 CreateType::Background => {
-                    tracing::error!(id = stream_job.id(), error = ?e, "finish stream job failed")
+                    tracing::error!(id = job_id, error = ?e, "finish stream job failed");
+                    if let Err(err) = self.fragment_manager.select_table_fragments_by_table_id(&job_id.into()).await
+                        && err.is_fragment_not_found() {
+                        // If the table fragments are not found, it means that the stream job has not been created.
+                        // We need to cancel the stream job.
+                        self.cancel_stream_job(&stream_job, internal_tables, Some(&e))
+                            .await?;
```
Comment (BugenZhao) [inline src/meta/src/rpc/ddl_controller.rs:860]: The `cancel` looks confusing to me. 🤣 It appears to be "cleanup".

File: src/meta/src/rpc/ddl_controller.rs [hunk 3]
```diff
@@ -850,10 +850,18 @@ impl DdlController {
...
+                    } else {
+                        // NOTE: This assumes that we will trigger recovery,
```
Comment (BugenZhao) [inline src/meta/src/rpc/ddl_controller.rs:862]: So in this branch, the job will be resumed after recovery, right?

**Review by BugenZhao — APPROVED** [review]
Really tricky. 🥵

File: src/meta/src/rpc/ddl_controller.rs (hunk 3 above)
Comment (yezizp2012) [reply src/meta/src/rpc/ddl_controller.rs:862]: Yes exactly.

File: src/meta/src/rpc/ddl_controller.rs (hunk 2 above)
Comment (yezizp2012) [reply src/meta/src/rpc/ddl_controller.rs:860]: NTFS. 🥵

**Pull request merged**